payments/models.py

```diff
@@ -154,3 +171,374 @@ def __str__(self):
+
+        # Check that the lease customer and the received customer are the same
+        if self.lease.customer != self.customer:
+            raise ValidationError(
+                _("The lease provided belongs to a different customer")
+            )
+
+    def _check_product_and_lease(self) -> None:
+        if isinstance(self.product, BerthProduct) and not isinstance(
+            self.lease, BerthLease
+        ):
+            raise ValidationError(
+                _("A BerthProduct must be associated with a BerthLease")
+            )
+        elif isinstance(self.product, WinterStorageProduct) and not isinstance(
+            self.lease, WinterStorageLease
+        ):
+            raise ValidationError(
+                _("A WinterStorageProduct must be associated with a WinterStorageLease")
+            )
+
+    def _check_product_or_price(self) -> None:
+        if all([not self.product, not self._product_object_id, not self.price]):
+            raise ValidationError(
+                _("Order must have either product object or price value")
+            )
+
+    def _check_product_and_price(self) -> None:
+        if (
+            self.price
+            and self.tax_percentage
+            and (self.product or self._product_object_id)
+        ):
+            raise ValidationError(
+                _("Cannot receive product and plain price/tax_percentage")
+            )
+
+    def _check_same_product(self, old_instance) -> None:
+        if old_instance.product != self.product:
+            raise ValidationError(_("Cannot change the product assigned to this order"))
+
+    def _check_same_lease(self, old_instance) -> None:
+        if old_instance.lease and self.lease != old_instance.lease:
+            raise ValidationError(
+                _("Cannot change the lease associated with this order")
+            )
+
+    def clean(self):
+        # Check that only Berth or Winter products are passed (not other models)
+        self._check_valid_products()
+
+        if self.lease:
+            self._check_valid_lease()
+
+            # Check that product and lease are from the same type
+            if self.product:
+                self._check_product_and_lease()
+
+        # Check that it has either product or price
+        self._check_product_or_price()
+
+        if self._state.adding:
+            # Don't allow creating with product and price
+            self._check_product_and_price()
+        else:
+            old_instance = Order.objects.get(id=self.id)
+            # If the product is being changed
+            self._check_same_product(old_instance)
+
+            # If the lease is being changed
+            self._check_same_lease(old_instance)
+
+    def is_full_season(self):
+        return (
+            self.lease.start_date == calculate_berth_lease_start_date()
+            and self.lease.end_date == calculate_berth_lease_end_date()
+        )
```

Comment:
this should probably handle no `self.lease` more gracefully...

Reply:
Hmmm good catch. We could either check it _before_ calling `is_full_season` or... I don't think returning `False` would make sense in this context. Raising an error maybe?

Reply:
hm, but it would already raise an `AttrivbuteError` :simple_smile: 🤔

Reply:
ok, let's assume that this check should always follow `if order.lease` 🙂

Reply:
Ok ok, checking for checks. 👍

Reply:
I mean, this function can maybe stay as is. 🙂

Reply:
Yups, I meant that I'll check that we check for `lease` before calling it. :derp: